### PR TITLE
Fix/suitesparse

### PR DESCRIPTION
**Pull request by gjbex**
Fixed two problems with this easyblock:
1) it assumes variable definitions in the .mk file will not have leading whitespace (assumption doesn't hold for SuiteSparse 4.4.3),
2) the mechanism for ensuring that all variables not replaced in the .mk file are appended works only if the variable name does not occur in any line (e.g., comment) prior to its definition (definitely not the case in SuiteSparse 4.4.3).

Although the current code works, it is not future proof.  In Python 2, the items() method returns a list of key/value tuples, hence it is save to apply the pop() method to the cfgvars dictionary.  In Python 3, the items() method will return an iterator, and removing items from a data structure one is iterating over is not a good idea.
I've only added a comment, since I assume that this pattern will be quite pervasive given the list of authors of this block, so it should probably be addressed more systematically.

### CODE REVIEW

**Comment by boegel** [comment]
Jenkins: ok to test

File: easybuild/easyblocks/s/suitesparse.py [hunk 1]
```diff
@@ -94,8 +94,12 @@ def configure_step(self):
...
             for line in fileinput.input(fp, inplace=1, backup='.orig'):
                 for (k, v) in cfgvars.items():
-                    line = re.sub(r"^(%s\s*=\s*).*$" % k, r"\1 %s # patched by EasyBuild" % v, line)
-                    if k in line:
+                    orig_line = line
+                    line = re.sub(r"^\s*(%s\s*=\s*).*$" % k,
+                                  r"\1 %s # patched by EasyBuild" % v,
+                                  line)
+                    if line != orig_line:
+# Note: this will crash & burn in Python 3
```
Comment (boegel) [inline easybuild/easyblocks/s/suitesparse.py]: there's lots of stuff all throughout EasyBuild right now which is not Python 3 compatible, so this comment has little value I think

also: the comment should be indented with the code it applies to

File: easybuild/easyblocks/s/suitesparse.py [hunk 2]
```diff
@@ -94,8 +94,12 @@ def configure_step(self):
         try:
             for line in fileinput.input(fp, inplace=1, backup='.orig'):
```
Comment (boegel) [inline easybuild/easyblocks/s/suitesparse.py]: let's move down the definition of `fp` closer to where it's actually being used, i.e. right above the `try`...

File: easybuild/easyblocks/s/suitesparse.py [hunk 3]
```diff
@@ -94,8 +94,12 @@ def configure_step(self):
...
                 for (k, v) in cfgvars.items():
-                    line = re.sub(r"^(%s\s*=\s*).*$" % k, r"\1 %s # patched by EasyBuild" % v, line)
-                    if k in line:
+                    orig_line = line
+                    line = re.sub(r"^\s*(%s\s*=\s*).*$" % k,
+                                  r"\1 %s # patched by EasyBuild" % v,
+                                  line)
```
Comment (boegel) [inline easybuild/easyblocks/s/suitesparse.py]: no need to split this across three lines?

File: easybuild/easyblocks/s/suitesparse.py [hunk 4]
```diff
@@ -94,8 +94,12 @@ def configure_step(self):
         try:
             for line in fileinput.input(fp, inplace=1, backup='.orig'):
                 for (k, v) in cfgvars.items():
-                    line = re.sub(r"^(%s\s*=\s*).*$" % k, r"\1 %s # patched by EasyBuild" % v, line)
-                    if k in line:
+                    orig_line = line
+                    line = re.sub(r"^\s*(%s\s*=\s*).*$" % k,
```
Comment (boegel) [inline easybuild/easyblocks/s/suitesparse.py]: include a comment explaining what's going on here?

File: easybuild/easyblocks/s/suitesparse.py [hunk 5]
```diff
@@ -94,8 +94,12 @@ def configure_step(self):
         try:
             for line in fileinput.input(fp, inplace=1, backup='.orig'):
                 for (k, v) in cfgvars.items():
```
Comment (boegel) [inline easybuild/easyblocks/s/suitesparse.py]: replace `k` and `v` by non-single-letter variable names (`key` and `val` will do, but more meaningful (yet still short) would be even better)

**Comment by boegel** [comment]
@gjbex: can you tweak the PR title a bit to try and summarise the patch in a single line? that will be a tremendous help in the `git log` that will contain `Merge pull request #541 from hpcleuven:fix/suitesparse: <good title goes here>`

File: easybuild/easyblocks/s/suitesparse.py (hunk 3 above)
Comment (gjbex) [inline easybuild/easyblocks/s/suitesparse.py]: When doing a substitution, I find it useful to be able to visually compare the two expressions, aligning them vertically helps avoiding bugs.

**Comment by gjbex** [comment]
Yes, I can tweak it, but allow me to make a few things clear.

1) Code reviews are definitely a good thing, sometimes even for the reviewers.
2) I don't mind issues with my code being raised, and  I certainly don't mind fixing them, they are my responsibility.
3) I do mind getting suggestions to fix someone else's code though, since it is not my responsibility.  I would suggest to take that up with the original authors since they may have had some good reason to do what they did.

File: easybuild/easyblocks/s/suitesparse.py (hunk 5 above)
Comment (gjbex) [inline easybuild/easyblocks/s/suitesparse.py]: Please be consistent in your code review, there's another one like it on on 109.

File: easybuild/easyblocks/s/suitesparse.py (hunk 4 above)
Comment (gjbex) [inline easybuild/easyblocks/s/suitesparse.py]: For a simple for-loop?

Comment (boegel) [inline easybuild/easyblocks/s/suitesparse.py]: No, for the regex substitution, maybe just make the existing `patch file` comment a bit clearer, to explain what the regex in the for loop does?

**Comment by gjbex** [comment]
Fixed all issues test report coming with PR 1328 for easyconfigs.

**Comment by boegel** [comment]
@gjbex: thanks a lot for tackling the remarks, despite your reservations, much appreciated

As discussed off-PR, I realise our way of giving feedback can be quite aggressive, and daunting for people new to the process.

In this particular case, you were unlucky to (having to) touch an easyblock that hasn't been getting love for quite some time (Feb'14), and hence is behind on the code style standard we're currently trying to maintain. The code style is constantly evolving (not necessarily changing much though), so it's important to try and resolve any issues we run into when they're being put in the spotlight.

That being said, I understand that you feel like you shouldn't be cleaning up other people's mess. Question is, if we don't tackle it when it's being thrown in our face, when will we ever...

I hope this doesn't put you off w.r.t. to future contributions, since you're clearly "on a roll" the last couple of weeks.

**Comment by boegel** [comment]
Tested this together with https://github.com/hpcugent/easybuild-easyconfigs/pull/1328, working great, looking good, so going in. Thanks again @gjbex!

**Pull request merged**